lxd/storage/backend_lxd_patches.go

```diff
@@ -1,17 +1,96 @@
 package storage
 
 import (
+	"github.com/pkg/errors"
+
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/project"
+	"github.com/lxc/lxd/lxd/revert"
 	"github.com/lxc/lxd/lxd/storage/drivers"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+	"github.com/lxc/lxd/shared/logger"
 )
 
 var lxdEarlyPatches = map[string]func(b *lxdBackend) error{}
 
 var lxdLatePatches = map[string]func(b *lxdBackend) error{
-	"storage_create_vm":       lxdPatchStorageCreateVM,
-	"storage_create_vm_again": lxdPatchStorageCreateVM,
+	"storage_create_vm":                        lxdPatchStorageCreateVM,
+	"storage_create_vm_again":                  lxdPatchStorageCreateVM,
+	"storage_rename_custom_volume_add_project": lxdPatchStorageRenameCustomVolumeAddProject,
 }
 
 // Patches start here.
 func lxdPatchStorageCreateVM(b *lxdBackend) error {
 	return b.createStorageStructure(drivers.GetPoolMountPath(b.name))
 }
+
+// lxdPatchStorageRenameCustomVolumeAddProject renames all custom volumes in the default project (which is all of
+// the custom volumes right now) to have the project prefix added to the storage device volume name.
+// This is so we can added project support to custom volumes and avoid any name collisions.
+func lxdPatchStorageRenameCustomVolumeAddProject(b *lxdBackend) error {
+	// At this time, all custom volumes are in the default project.
+	volumes, err := b.state.Cluster.StoragePoolVolumesGet(project.Default, b.ID(), []int{db.StoragePoolVolumeTypeCustom})
+	if err != nil && err != db.ErrNoSuchObject {
+		return errors.Wrapf(err, "Failed getting custom volumes for default project")
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	for _, v := range volumes {
+		// Run inside temporary function to ensure revert has correct volume scope.
+		err = func(curVol *api.StorageVolume) error {
+			// There's no need to pass the config as it's not needed when renaming a volume.
+			oldVol := b.newVolume(drivers.VolumeTypeCustom, drivers.ContentTypeFS, curVol.Name, nil)
+
+			// Add default project prefix to current volume name.
+			newVolStorageName := project.StorageVolume(project.Default, curVol.Name)
+
+			// Check if volume is currently mounted.
+			oldMntPath := drivers.GetVolumeMountPath(b.Name(), drivers.VolumeTypeCustom, curVol.Name)
+
+			// If the volume is mounted we need to be careful how we rename it to avoid interrupting a
+			// running instance's attached volumes.
+			ourUnmount := false
+			if shared.IsMountPoint(oldMntPath) {
+				logger.Infof("Lazy unmount custom volume %q in pool %q", curVol.Name, b.Name())
+				_, err = shared.RunCommand("umount", "--lazy", oldMntPath)
```

Comment:
We usually just call unix.Unmount with MNT_DETACH instead.

Reply:
Have fixed.